server: pod-to-pod fanout for statements api on tenant

dhartunian opened the pull request:
Previously, the Statements endpoint implementation was purely local in
its execution and reported only in-memory data on the current tenant.
This change initializes a gRPC and gRPC-gateway server on the tenant,
much like we do on the hosts already, and uses the newly added
InstanceID subsystem for tenants to implement a gRPC-based fanout for
the Statements endpoint implementation.

The fanout is done much in the same manner as on hosts and we
expose the status server endpoints via HTTP on the tenant as well.

This change is necessary to support serverless observability features
and provide our UIs access to the Statements endpoint. Future work may
move this API to SQL-only

Resolves cockroachdb#64477

Release note (api change): tenant pods now expose the Statements API at
`/_status/statements` on their HTTP port.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/66675)
<!-- Reviewable:end -->

rimadeodhar reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66675)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @knz)

___
*[pkg/server/statements.go, line 57 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-McpnFPH0aYP2vzxZ8tX:-McpnFPH0aYP2vzxZ8tY:byrvo37) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef1c947e14d7681202077f15dda23bf034553c7b/pkg/server/statements.go#L57)):*
> ```Go
> 			if local {
> 				// TODO(davidh): putting an InstanceID in a NodeIDContainer...can this be avoided?
> 				container := &base.NodeIDContainer{}
> ```

Is SqlIDContainer a better fit here: https://github.com/cockroachdb/cockroach/blob/e53b17af66f158e5ba6412b6913b9b67616f633b/pkg/base/node_id.go#L98? I think the function statementsLocal will likely need some changes too though.
___
*[pkg/server/statements.go, line 89 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-Mcpo0Q8C-U4F7PHddcg:-Mcpo0Q8C-U4F7PHddch:bwqyg2v) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef1c947e14d7681202077f15dda23bf034553c7b/pkg/server/statements.go#L89)):*
> ```Go
> 			},
> 			func(instanceID base.SQLInstanceID, err error) {
> 				// TODO(couchand): do something here...
> ```

What should happen here?
___
*[pkg/server/status.go, line 1965 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-McprKWcEmEtoCBqqglx:-McprKWcEmEtoCBqqgly:b-bdz4cr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef1c947e14d7681202077f15dda23bf034553c7b/pkg/server/status.go#L1965)):*
> ```Go
> 	responseChan := make(chan instanceResponse, numInstances)
> 
> 	// TODO(davidh): somehow the tenantInstances above must either have a liveness state
> ```

The GetAllInstancesForTenant will only return the active instances at that point. What do we need this liveness state for in addition to that?


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> - Some of this code uses a `NodeID` as an instance ID to keep the
>  API from splitting too much. Is it important to avoid this? The
>  types both alias the same int32 type.

See my point on the linked PR  #66506 - I'd say let's unify both types and avoid duplicating the code.

> - The `iteratePods` implementation is copied over mostly from the
>  `iterateNodes` function. Should efforts be taken to support both
>  options in a single function call? I held off on that until we know
>  more about how much the working implementations will differ.

I would say make 1 function that does both, and then make `iteratePods` and `iterateNode` thin wrappers that call the common function with a flag to disambiguate.

It's also possible that once/if we unify node and instance ID types, the only place where we need to choose a different behavior is the logic that extracts the remote address to connect, and the remainder of the code remains the same?

> Reviewed 14 of 14 files at r1, 2 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-:-MdH-e85EpVfI6cBxTJ7:b-gklslo)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @knz)

___
*[pkg/server/testserver.go, line 601 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MdGz0yN01_7Y0Iwiyjf:-MdGz0yO5z_hj_yZ2fDi:bqgmt0f) ([raw file](https://github.com/cockroachdb/cockroach/blob/eefeb2a5e4b1ab10b2a5eb97c6b6ba505c62315c/pkg/server/testserver.go#L601)):*
> ```Go
> 	contentionRegistry := contention.NewRegistry()
> 	flowScheduler := flowinfra.NewFlowScheduler(baseCfg.AmbientCtx, stopper, st)
> 	codec := keys.MakeSQLCodec(sqlCfg.TenantID)
> ```

Can you extract this code into a shared function used both here and in server.go.


<!-- Sent from Reviewable.io -->

dhartunian committed:
server: pod-to-pod fanout for statements api on tenant

Previously, the Statements endpoint implementation was purely local in
its execution and reported only in-memory data on the current tenant.
This change initializes a gRPC and gRPC-gateway server on the tenant,
much like we do on the hosts already, and uses the newly added
InstanceID subsystem for tenants to implement a gRPC-based fanout for
the Statements endpoint implementation.

The fanout is done much in the same manner as on hosts and we
expose the status server endpoints via HTTP on the tenant as well.

This change is necessary to support serverless observability features
and provide our UIs access to the Statements endpoint. Future work may
move this API to SQL-only

Resolves #64477

Release note (api change): tenant pods now expose the Statements API at
`/_status/statements` on their HTTP port.

dhartunian reviewed (commented):
> See my point on the linked PR #66506 - I'd say let's unify both types and avoid duplicating the code.

I'm reticent to take this road because the current code makes it very clear when we cast an InstanceID to a NodeID when we want to use existing code. My worry is that if they're easily interchangeable it can cause a lot more confusion when we misuse one in place of the other. We certainly don't want to want to support InstanceIDs in all places where a NodeID is accepted.

> I would say make 1 function that does both, and then make iteratePods and iterateNode thin wrappers that call the common function with a flag to disambiguate.

I really wanted this to work and gave it an attempt but found that there was too much difference between the two implementations to easily unify. All the callbacks passed into `iterateX` take different types due to the InstanceID and NodeID difference and we have different ways of processing the node and instance lists from the liveness vs instanceID subsystems. I can give it another shot if you think one piece of code that's more complex is worth it but I wonder if the near-term goal is to remove the `tenant_status.go` implementation anyway, that we can keep as-is until that happens.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-:-MfEed84CYhgM7hNsTbJ:b-9gdv3g)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @dt, and @knz)



<!-- Sent from Reviewable.io -->

andy-kimball commented:
Is this change back-portable to 21.1, or does it only work on top of 21.2? If it only works on 21.2, it would force us to deliver the initial release of Serverless on a beta of 21.2, which we've been trying to avoid having to do.





<!-- Sent from Reviewable.io -->

knz reviewed (commented):
I have reviewed the PR and it's generally OK assuming the instance ID subsystem works. I have an important question about authorization (see below) but I think this can be fixed by using existing API calls.

> Is this change back-portable to 21.1

@andy-kimball we definitely should talk about this. For now the PR as-is depends on #66600  which is very large and risky to backport. What I'd like to suggest to de-risk this is to keep David's PR (this one) but rebase it on top of a simpler change which provides the instance ID of SQL pods on the command line, as a command-line argument. This would be a transient solution which would be backportable, and would give us the time to let Rima's PR bake on the 22.1 branch until we iron its bugs out.

> Reviewed 56 of 56 files at r3, 7 of 11 files at r4, 17 of 19 files at r5, 10 of 10 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-:-MfIN6rAA580RHv_rr8q:b5leu4k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @dt)

___
*[pkg/rpc/auth_tenant.go, line 58 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfIM2HkA-twWeq19Cdt:-MfIM2HkA-twWeq19Cdu:b-5gpz6p) ([raw file](https://github.com/cockroachdb/cockroach/blob/11febcbd36a177ba89e7ff9668e2fee4681af785/pkg/rpc/auth_tenant.go#L58)):*
> ```Go
> 		return nil // no authorization
> 
> 	case "/cockroach.server.serverpb.Status/Statements":
> ```

what is going on here? How come is this not subject to authorization?
___
*[pkg/server/tenant_status.go, line 168 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfIMcyI3EVafXl16WNt:-MfIMcyI3EVafXl16WNu:b-lq94o2) ([raw file](https://github.com/cockroachdb/cockroach/blob/11febcbd36a177ba89e7ff9668e2fee4681af785/pkg/server/tenant_status.go#L168)):*
> ```Go
> func (t *tenantStatusServer) Statements(
> 	ctx context.Context, req *serverpb.StatementsRequest,
> ) (*serverpb.StatementsResponse, error) {
> ```

if you use `propagateGatewayMetadata` here, you can forward authn/authz from the outside to the fan-out. We can chat about this.
___
*[pkg/server/tenant_status.go, line 193 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfIMQ6KDLLP96W8pT6E:-MfIMQ6KDLLP96W8pT6F:bl8cxn2) ([raw file](https://github.com/cockroachdb/cockroach/blob/11febcbd36a177ba89e7ff9668e2fee4681af785/pkg/server/tenant_status.go#L193)):*
> ```Go
> 			container := base.NodeIDContainer{}
> 			container.Set(ctx, roachpb.NodeID(t.sqlServer.SQLInstanceID()))
> 			return statementsLocal(ctx, &container, t.sqlServer)
> ```

I am confused, why does this need a node ID container and not the node ID directly?
___
*[pkg/util/log/clog.go, line 227 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfILvN5D9WNQ23w30F2:-MfILvN5D9WNQ23w30F3:bs7a60k) ([raw file](https://github.com/cockroachdb/cockroach/blob/11febcbd36a177ba89e7ff9668e2fee4681af785/pkg/util/log/clog.go#L227)):*
> ```Go
> 	defer logging.idMu.Unlock()
> 
> 	// TODO(davidh): PRIOR TO MERGE: since my test in `tenant_grpc_test.go` calls
> ```

Use `log.TestingClearServerIdentifiers()` in your tests. See `pkg/ccl/backup/backup_test.go` for some example uses.


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
> For now the PR as-is depends on #66600 which is very large and risky to backport.

Our intention with this PR and #66600 was to backport it all along. I understand there's a lot of code in the pre-requisite PR but if most of the code paths are triggered by multi-tenant doesn't that de-risk it? (cc @rimadeodhar)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-:-MfIosAE6pgSoCpy8fAt:bpg3xcp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @dt)



<!-- Sent from Reviewable.io -->

Azhng reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66675)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @dt)

___
*[pkg/server/tenant_status.go, line 151 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfZ1-rk3ie5puZt4Gv1:-MfZ1-rk3ie5puZt4Gv2:b-r8l5ow) ([raw file](https://github.com/cockroachdb/cockroach/blob/11febcbd36a177ba89e7ff9668e2fee4681af785/pkg/server/tenant_status.go#L151)):*
> ```Go
> }
> 
> func (t *tenantStatusServer) ResetSQLStats(
> ```

this also needs to be updated to perform RPC fanout. 
___
*[pkg/server/tenant_status.go, line 160 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfZ239y-EVUR54hpdyT:-MfZ239z67ODJCia-J_v:bay9z0d) ([raw file](https://github.com/cockroachdb/cockroach/blob/11febcbd36a177ba89e7ff9668e2fee4681af785/pkg/server/tenant_status.go#L160)):*
> ```Go
> // Statements implements the relevant endpoint on the StatusServer by
> // fanning out a request to all pods on the current tenant via gRPC to collect
> // in-memory statistics and aggregate them for the caller.
> ```

nit: I think it's slightly misleading here, we are only trivially append stats from other nodes into a slice but I don't think we aggregate them. 
___
*[pkg/server/tenant_status.go, line 184 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfZ4noXDnIdqVR4ulxH:-MfZ4noXDnIdqVR4ulxI:b-kn62l3) ([raw file](https://github.com/cockroachdb/cockroach/blob/11febcbd36a177ba89e7ff9668e2fee4681af785/pkg/server/tenant_status.go#L184)):*
> ```Go
> 
> 	if len(req.NodeID) > 0 {
> 		// We are interpreting the `NodeID` in the request as an `InstanceID` since
> ```

This reinterpretation cast feels a bit strange to me. Preferably maybe we should consider either: 
* rename the `NodeID` field, Or 
* introduce another `InstanceID` field, Or
* introduce a `oneof` field to include both `NodeID` and `InstanceID` 
___
*[pkg/server/tenant_status.go, line 195 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfZ6Kxs8YkyT8ys1YUz:-MfZ6Kxs8YkyT8ys1YV-:b-cug7v2) ([raw file](https://github.com/cockroachdb/cockroach/blob/11febcbd36a177ba89e7ff9668e2fee4681af785/pkg/server/tenant_status.go#L195)):*
> ```Go
> 			return statementsLocal(ctx, &container, t.sqlServer)
> 		}
> 		// Iterate through live instances to find the one we want to target
> ```

Is this iteration necessary? Can we have the `InstanceID` subsystem expose a lookup API instead? 
___
*[pkg/server/tenant_status.go, line 249 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfZ6dN304wtoGwuwyK_:-MfZ6dN304wtoGwuwyKa:b2woqgy) ([raw file](https://github.com/cockroachdb/cockroach/blob/11febcbd36a177ba89e7ff9668e2fee4681af785/pkg/server/tenant_status.go#L249)):*
> ```Go
> func (t *tenantStatusServer) parseInstanceID(
> 	instanceIDParam string,
> ) (base.SQLInstanceID, bool, error) {
> ```

nit: Can we name the return values here. It's not exactly clear what the second `bool` is


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66675)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @dt, @knz, and @rimadeodhar)

___
*[pkg/rpc/auth_tenant.go, line 58 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfIM2HkA-twWeq19Cdt:-MfYMsaDCAjHkffLTgAf:bam1lbh) ([raw file](https://github.com/cockroachdb/cockroach/blob/11febcbd36a177ba89e7ff9668e2fee4681af785/pkg/rpc/auth_tenant.go#L58)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

what is going on here? How come is this not subject to authorization?
</blockquote></details>

I was a bit confused about this method's purpose. The other endpoints seemed to utilize KV APIs instead of SQL which is why they had special authorization requirements.

The statements endpoint only uses SQL and has an auth chunk at the top of its implementation so I assumed that would be adequate. I can add a comment to that effect here unless you think I'm misunderstanding.
___
*[pkg/server/status.go, line 1965 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-McprKWcEmEtoCBqqglx:-MfYqxwx3u7AqePBEhqa:b-a70ypz) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef1c947e14d7681202077f15dda23bf034553c7b/pkg/server/status.go#L1965)):*
<details><summary><i>Previously, rimadeodhar (Rima Deodhar) wrote…</i></summary><blockquote>

The GetAllInstancesForTenant will only return the active instances at that point. What do we need this liveness state for in addition to that?
</blockquote></details>

It was just because the errors contained some liveness state info. It's been omitted. And I've removed the `isAlive` filtering I had elsewhere in here since you've added that functionality.
___
*[pkg/server/tenant_status.go, line 151 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfZ1-rk3ie5puZt4Gv1:-MfZqtQk4a0plGqj5jSQ:bt9klg6) ([raw file](https://github.com/cockroachdb/cockroach/blob/11febcbd36a177ba89e7ff9668e2fee4681af785/pkg/server/tenant_status.go#L151)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

this also needs to be updated to perform RPC fanout. 
</blockquote></details>

I'll make a separate ticket for that, don't want to make this PR too big. Is that the only other endpoint we'll need to support? Might ask your team to pick that up if y'all have the capacity.
___
*[pkg/server/tenant_status.go, line 160 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfZ239y-EVUR54hpdyT:-MfZqwvACHKo20G-vldV:b-693y4w) ([raw file](https://github.com/cockroachdb/cockroach/blob/11febcbd36a177ba89e7ff9668e2fee4681af785/pkg/server/tenant_status.go#L160)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

nit: I think it's slightly misleading here, we are only trivially append stats from other nodes into a slice but I don't think we aggregate them. 
</blockquote></details>

Good point. Updated message.
___
*[pkg/server/tenant_status.go, line 168 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfIMcyI3EVafXl16WNt:-MfYsVW81KUQAktg1c2Z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/11febcbd36a177ba89e7ff9668e2fee4681af785/pkg/server/tenant_status.go#L168)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

if you use `propagateGatewayMetadata` here, you can forward authn/authz from the outside to the fan-out. We can chat about this.
</blockquote></details>

Done.
___
*[pkg/server/tenant_status.go, line 184 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfZ4noXDnIdqVR4ulxH:-MfZvImz5N0WpUDtL1Q7:b63ogzg) ([raw file](https://github.com/cockroachdb/cockroach/blob/11febcbd36a177ba89e7ff9668e2fee4681af785/pkg/server/tenant_status.go#L184)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

This reinterpretation cast feels a bit strange to me. Preferably maybe we should consider either: 
* rename the `NodeID` field, Or 
* introduce another `InstanceID` field, Or
* introduce a `oneof` field to include both `NodeID` and `InstanceID` 

</blockquote></details>

I did consider this and was wondering if it was really worth the hassle of modifying the API. Since tenants don't have a `NodeID` it seems clear that the instanceID makes sense in that context. The `NodeID` field is used in the responses of these endpoints quite extensively and I hesitate to either make sweeping changes or introduce inconsistency by only changing the request type.
___
*[pkg/server/tenant_status.go, line 193 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfIMQ6KDLLP96W8pT6E:-MfYst9R9uLahL316b4X:b-8oo5v4) ([raw file](https://github.com/cockroachdb/cockroach/blob/11febcbd36a177ba89e7ff9668e2fee4681af785/pkg/server/tenant_status.go#L193)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I am confused, why does this need a node ID container and not the node ID directly?
</blockquote></details>

Great question. I modified `statementsLocal` to just take the NodeID directly. It was just unwrapping it anyway.
___
*[pkg/server/tenant_status.go, line 195 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfZ6Kxs8YkyT8ys1YUz:-MfZxt9Y4eB-p2Bu0QED:bq9kt5y) ([raw file](https://github.com/cockroachdb/cockroach/blob/11febcbd36a177ba89e7ff9668e2fee4681af785/pkg/server/tenant_status.go#L195)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Is this iteration necessary? Can we have the `InstanceID` subsystem expose a lookup API instead? 
</blockquote></details>

It does! Good suggestion. I modified the `dialPod` interface to support separate address lookups so we can supply them via direct lookup or from the instanceInfo when we have it.
___
*[pkg/server/tenant_status.go, line 249 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfZ6dN304wtoGwuwyK_:-MfZy6loESrK8PneNaZG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/11febcbd36a177ba89e7ff9668e2fee4681af785/pkg/server/tenant_status.go#L249)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

nit: Can we name the return values here. It's not exactly clear what the second `bool` is
</blockquote></details>

Done.
___
*[pkg/server/testserver.go, line 601 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MdGz0yN01_7Y0Iwiyjf:-MfYt32dE5GHeDImn0eP:blrsfpq) ([raw file](https://github.com/cockroachdb/cockroach/blob/eefeb2a5e4b1ab10b2a5eb97c6b6ba505c62315c/pkg/server/testserver.go#L601)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Can you extract this code into a shared function used both here and in server.go.
</blockquote></details>

cc @rimadeodhar these changes are part of your PR.
___
*[pkg/util/log/clog.go, line 227 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfILvN5D9WNQ23w30F2:-MfZqogkEb58oKmp9Wqh:b-4659eo) ([raw file](https://github.com/cockroachdb/cockroach/blob/11febcbd36a177ba89e7ff9668e2fee4681af785/pkg/util/log/clog.go#L227)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Use `log.TestingClearServerIdentifiers()` in your tests. See `pkg/ccl/backup/backup_test.go` for some example uses.
</blockquote></details>

Done. Thx for the tip
___
*[pkg/server/statements.go, line 57 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-McpnFPH0aYP2vzxZ8tX:-MfYtFbN9c8_mXeBCyT9:bcom7lx) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef1c947e14d7681202077f15dda23bf034553c7b/pkg/server/statements.go#L57)):*
<details><summary><i>Previously, rimadeodhar (Rima Deodhar) wrote…</i></summary><blockquote>

Is SqlIDContainer a better fit here: https://github.com/cockroachdb/cockroach/blob/e53b17af66f158e5ba6412b6913b9b67616f633b/pkg/base/node_id.go#L98? I think the function statementsLocal will likely need some changes too though.
</blockquote></details>

I ended up modifying `statementsLocal` to take the NodeID directly so just a basic conversion is enough.
___
*[pkg/server/statements.go, line 89 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-Mcpo0Q8C-U4F7PHddcg:-MfYtCes9DCezKEA3jNE:b-gyu4jk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef1c947e14d7681202077f15dda23bf034553c7b/pkg/server/statements.go#L89)):*
<details><summary><i>Previously, rimadeodhar (Rima Deodhar) wrote…</i></summary><blockquote>

What should happen here?
</blockquote></details>

I added logging for the errors.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> Reviewed 17 of 59 files at r7, 6 of 20 files at r8, 18 of 31 files at r9, 22 of 22 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66675)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @dhartunian, @dt, and @rimadeodhar)

___
*[pkg/rpc/auth_tenant.go, line 58 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfIM2HkA-twWeq19Cdt:-Mfcg9-AE04Y7TfcdH8_:bly34jn) ([raw file](https://github.com/cockroachdb/cockroach/blob/11febcbd36a177ba89e7ff9668e2fee4681af785/pkg/rpc/auth_tenant.go#L58)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

I was a bit confused about this method's purpose. The other endpoints seemed to utilize KV APIs instead of SQL which is why they had special authorization requirements.

The statements endpoint only uses SQL and has an auth chunk at the top of its implementation so I assumed that would be adequate. I can add a comment to that effect here unless you think I'm misunderstanding.
</blockquote></details>

My question is why did you modify this file at all.
I don't understand why this change was needed. What does this change do? What is prevented if you do not change this?
___
*[pkg/server/statements.go, line 51 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfcgaIABw_C0_ArEdXY:-MfcgaIABw_C0_ArEdXZ:bihw8kq) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad724ac13a3c02a6abc5492c79f41feca27c33ae/pkg/server/statements.go#L51)):*
> ```Go
> 		}
> 		if local {
> 			return statementsLocal(ctx, s.gossip.NodeID.Get(), s.admin.server.sqlServer)
> ```

Somewhere at the top of this function :if the node ID (instance ID) is not known yet, make the endpoint return an error that the service is not ready.


<!-- Sent from Reviewable.io -->

Azhng reviewed (commented):
> Reviewed 17 of 22 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66675)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @dt, @knz, and @rimadeodhar)

___
*[pkg/server/tenant_status.go, line 151 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfZ1-rk3ie5puZt4Gv1:-MfdAqh_A7j5e_KV3cXm:bb6mt9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/11febcbd36a177ba89e7ff9668e2fee4681af785/pkg/server/tenant_status.go#L151)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

I'll make a separate ticket for that, don't want to make this PR too big. Is that the only other endpoint we'll need to support? Might ask your team to pick that up if y'all have the capacity.
</blockquote></details>

:+1:
___
*[pkg/server/tenant_status.go, line 184 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfZ4noXDnIdqVR4ulxH:-MfdNhdiBgFSf_PHIGwm:bnyo7ej) ([raw file](https://github.com/cockroachdb/cockroach/blob/11febcbd36a177ba89e7ff9668e2fee4681af785/pkg/server/tenant_status.go#L184)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

I did consider this and was wondering if it was really worth the hassle of modifying the API. Since tenants don't have a `NodeID` it seems clear that the instanceID makes sense in that context. The `NodeID` field is used in the responses of these endpoints quite extensively and I hesitate to either make sweeping changes or introduce inconsistency by only changing the request type.
</blockquote></details>

Hmm perhaps we can leave a `C-cleanup` issue so we can clean this up later? For now I think at least we should update the comments in the `status.proto` on the multiple use of the `NodeID`. 

My concern is that for people without knowledge and context on history the `InstanceID` and `NodeID` this is something not entirely obvious. 

cc @knz , do you think this cleanup would be necessary? 
___
*[pkg/server/tenant_status.go, line 245 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfdKVwF1OQ4lnmTJi-L:-MfdKVwF1OQ4lnmTJi-M:b3n9sam) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad724ac13a3c02a6abc5492c79f41feca27c33ae/pkg/server/tenant_status.go#L245)):*
> ```Go
> 	// No parameter provided or set to local.
> 	if len(instanceIDParam) == 0 || localRE.MatchString(instanceIDParam) {
> 		return t.sqlServer.SQLInstanceID(), true, nil
> ```

super nit: would be great if you can add inline comments for folks that are not using Goland
e.g.

```
return t.sqlServer.SQLInstanceID(), true /* isLocal */, nil /* err */
```


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66675)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @dhartunian, @dt, and @rimadeodhar)

___
*[pkg/server/tenant_status.go, line 184 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfZ4noXDnIdqVR4ulxH:-MfgNjha1i_rqzL8X6Go:b-nrkxc4) ([raw file](https://github.com/cockroachdb/cockroach/blob/11febcbd36a177ba89e7ff9668e2fee4681af785/pkg/server/tenant_status.go#L184)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Hmm perhaps we can leave a `C-cleanup` issue so we can clean this up later? For now I think at least we should update the comments in the `status.proto` on the multiple use of the `NodeID`. 

My concern is that for people without knowledge and context on history the `InstanceID` and `NodeID` this is something not entirely obvious. 

cc @knz , do you think this cleanup would be necessary? 


</blockquote></details>

If we persist to keep a semantic difference between node ID and SQL instance ID (which I don't like, but I can be convinced to agree to disagree), then it's completely non-sensical to make the RPC service definition in the .proto file use the name "node ID". This is a pure SQL service and thus should only work with instance IDs.

(Moreover, in that line of thinking, I think that even non-multitenant clusters should define a SQL instance ID for every node, and we should clarify this throughout the code)

Unfortunately, we can't update the names of fields in .proto files because that changes the JSON interface. :sob: so we're stuck with that name for the time being.

I don't know how to think about this, however I agree a cleanup is in order.


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66675)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @dt, @knz, @rimadeodhar, and @stevendanna)

___
*[pkg/rpc/auth_tenant.go, line 58 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfIM2HkA-twWeq19Cdt:-MfhibA-BMitY3TbEeOj:boqa2e2) ([raw file](https://github.com/cockroachdb/cockroach/blob/11febcbd36a177ba89e7ff9668e2fee4681af785/pkg/rpc/auth_tenant.go#L58)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

My question is why did you modify this file at all.
I don't understand why this change was needed. What does this change do? What is prevented if you do not change this?
</blockquote></details>

If I remove that line the auth runs as an interceptor on all gRPC requests and causes the following error:

```
=== RUN   TestTenantGRPCServices/gRPC_is_running
    tenant_grpc_test.go:61: 
        	Error Trace:	tenant_grpc_test.go:61
        	Error:      	Received unexpected error:
        	            	rpc error: code = Unauthenticated desc = unknown method "/cockroach.server.serverpb.Status/Statements"
        	Test:       	TestTenantGRPCServices/gRPC_is_running
    --- FAIL: TestTenantGRPCServices/gRPC_is_running (0.01s)
```
___
*[pkg/server/statements.go, line 51 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfcgaIABw_C0_ArEdXY:-MfhdB9Z0cpxk1tQ93f3:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad724ac13a3c02a6abc5492c79f41feca27c33ae/pkg/server/statements.go#L51)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Somewhere at the top of this function :if the node ID (instance ID) is not known yet, make the endpoint return an error that the service is not ready.
</blockquote></details>

Done
___
*[pkg/server/tenant_status.go, line 151 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfZ1-rk3ie5puZt4Gv1:-Mfhd_hm2i-3k8A6zFnd:bhq6u74) ([raw file](https://github.com/cockroachdb/cockroach/blob/11febcbd36a177ba89e7ff9668e2fee4681af785/pkg/server/tenant_status.go#L151)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

:+1:
</blockquote></details>

Tracking here: https://github.com/cockroachdb/cockroach/issues/68177
___
*[pkg/server/tenant_status.go, line 184 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfZ4noXDnIdqVR4ulxH:-Mfhh6Uq7Ab9vWuO2N7d:b-t5nwzm) ([raw file](https://github.com/cockroachdb/cockroach/blob/11febcbd36a177ba89e7ff9668e2fee4681af785/pkg/server/tenant_status.go#L184)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

If we persist to keep a semantic difference between node ID and SQL instance ID (which I don't like, but I can be convinced to agree to disagree), then it's completely non-sensical to make the RPC service definition in the .proto file use the name "node ID". This is a pure SQL service and thus should only work with instance IDs.

(Moreover, in that line of thinking, I think that even non-multitenant clusters should define a SQL instance ID for every node, and we should clarify this throughout the code)

Unfortunately, we can't update the names of fields in .proto files because that changes the JSON interface. :sob: so we're stuck with that name for the time being.

I don't know how to think about this, however I agree a cleanup is in order.


</blockquote></details>

Added docs to the protobuf.

Cleanup ticket: https://github.com/cockroachdb/cockroach/issues/68178
___
*[pkg/server/tenant_status.go, line 245 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfdKVwF1OQ4lnmTJi-L:-MfhhiPy4p6pBfVPem37:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad724ac13a3c02a6abc5492c79f41feca27c33ae/pkg/server/tenant_status.go#L245)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

super nit: would be great if you can add inline comments for folks that are not using Goland
e.g.

```
return t.sqlServer.SQLInstanceID(), true /* isLocal */, nil /* err */
```
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> Reviewed 4 of 5 files at r11.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66675)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @dt, @rimadeodhar, and @stevendanna)

___
*[pkg/rpc/auth_tenant.go, line 58 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfIM2HkA-twWeq19Cdt:-Mfhx8aj5g8rtaim0oyQ:bfylbfi) ([raw file](https://github.com/cockroachdb/cockroach/blob/11febcbd36a177ba89e7ff9668e2fee4681af785/pkg/rpc/auth_tenant.go#L58)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

If I remove that line the auth runs as an interceptor on all gRPC requests and causes the following error:

```
=== RUN   TestTenantGRPCServices/gRPC_is_running
    tenant_grpc_test.go:61: 
        	Error Trace:	tenant_grpc_test.go:61
        	Error:      	Received unexpected error:
        	            	rpc error: code = Unauthenticated desc = unknown method "/cockroach.server.serverpb.Status/Statements"
        	Test:       	TestTenantGRPCServices/gRPC_is_running
    --- FAIL: TestTenantGRPCServices/gRPC_is_running (0.01s)
```
</blockquote></details>

ok now I understand better. thanks for checking.

1) I think there is still a misunderstanding though. Something that doesn't seem clear to you yet, and wasn't to me until I just checked, is that we have 3 different RPC servers to consider:

   (A) the RPC server run by a KV node for use by other KV nodes, and by client connections to the system tenant. This is also the common case of a non-multitenant regular node.
   (B) the RPC server run by a KV node for use by SQL pod processes.
   (C) the RPC server run by a SQL pod process.

   The case that interests us today is case (C).

   However, the function `authorize` we're looking at here is *only* relevant for case (B).  The struct `tenantAuthorizer`, as per its comment above in the same file, is for *"RPCs sent by tenants to a node's tenant RPC"*.   

   Therefore, the fact that there is an error when accessing the Statement endpoint using interface (B) is *expected* (and desirable!)

   There are two questions we should answer before moving forward here:

   - the error you observe suggests that a Statement RPC request is being sent in the (B) case somewhere in test `TestTenantGRPCServices`.  *That request should not happen*  - **the test is doing something invalid**. The problem is in that test and the fix is to change the test. How to prevent this test from sending this invalid request?

   - you will need a new test that sends a Statements request in both cases (A) and case (C), and check that both succeed. What would be a good location to implement this test?

2)  What was currently unclear around this code (and both you and I got confused) is that this entire function `authorize` is an *additional* authorization layer *on top* of the authn/authz mechanisms already implemented for all RPC endpoints, that's specific to case (B). Maybe this can be clarified (e.g. via a new comment on top of it).

3) Once the above is clarified, the comment `// no authorization` for the Ping request immediately above becomes more clearly nonsensical.  It is not the case that this endpoint is not subject to authorization just because this comment says so. The truth is that *the authorize method has nothing to add about it* and it is handled  in the common code path elsewhere. So the phrasing `// no authorization` is really confusing and should probably be phrased as `// no restriction to usage of this endpoint by tenants`. Ditto for "regions.


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66675)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @dt, @knz, @rimadeodhar, and @stevendanna)

___
*[pkg/rpc/auth_tenant.go, line 58 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfIM2HkA-twWeq19Cdt:-Mg7E43rBaS9ygAKt9Sa:b-pr4yro) ([raw file](https://github.com/cockroachdb/cockroach/blob/11febcbd36a177ba89e7ff9668e2fee4681af785/pkg/rpc/auth_tenant.go#L58)):*
> the error you observe suggests that a Statement RPC request is being sent in the (B) case somewhere in test TestTenantGRPCServices.  That request should not happen  - the test is doing something invalid. The problem is in that test and the fix is to change the test. How to prevent this test from sending this invalid request?

I don't think this is a test issue. The test dials the address of the tenant so it can't be accessing a (B) server. The issue is that tenant gRCP servers are configured in an identical manner to KV nodes. All gRPC servers are created using the same constructor `rpc.NewServer()` which adds these new auth interceptors for all requests that come from a tenant. When a tenant starts a gRPC server and accepts requests from other tenants, it opts-in in these authz routines as well.

Here are my options:

1. Make the `NewServer()` either configurable or aware if its being configured on a tenant node and disable the additional authz interceptors in those cases.
2. Make `authorize()` do something useful for tenant-tenant comms. It could, for instance, check that the client's tenantID matches the server tenantID. I don't think we explicitly check for that anywhere today.
3. Add a `// TODO` for more complex authz specification for tenant/tenant comms and move on

I'm taking option 2 as it provides us a way to retain common configuration while also restricting certain RPC calls to remain strictly between tenants when the client issuing the request is a tenant.

---

1. I've added a test `statements_test.go` in the `server` package that tests the (A) case. 

I added a 2 more test cases in my test file that issue a Statements RPC call (as opposed to an HTTP one) using one tenant's RPCContext to another (expecting success) and a request from tenant to KV (expecting failure)

2. Added comments
3. Added comments


<!-- Sent from Reviewable.io -->

rimadeodhar reviewed (commented):
LG on the whole, just had some minor nits and comments. Please rebase now the #66600 has been merged and I'll take a look again after.

> Reviewed 3 of 4 files at r12.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-:-MgHpmuo3wbkGKBhPXYG:be4cajw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @dhartunian, @dt, @knz, @rimadeodhar, and @stevendanna)

___
*[docs/generated/http/full.md, line 2803 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MgHmlFiDGV_SwXkzyw0:-MgHmlFiDGV_SwXkzyw1:b-fschzu) ([raw file](https://github.com/cockroachdb/cockroach/blob/73090bd82ead796cdd220762b842f7338d38142c/docs/generated/http/full.md#L2803)):*
> ```Markdown
> 
> StatementsRequest is used by both tenant and node-level
> implementations to server fan-out requests across multiple nodes or
> ```

Nit: Is this a typo, should it be serve fan-out requests?
___
*[pkg/ccl/serverccl/tenant_grpc_test.go, line 32 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MgHnEok3i9VFAkgk9Ww:-MgHnEok3i9VFAkgk9Wx:bpyy16f) ([raw file](https://github.com/cockroachdb/cockroach/blob/73090bd82ead796cdd220762b842f7338d38142c/pkg/ccl/serverccl/tenant_grpc_test.go#L32)):*
> ```Go
> // facing have been started up on the tenant server. This includes gRPC that is
> // used for pod-to-pod communication as well as the HTTP services powered by
> // gRPC Gateway that are used to serve e ndpoints to power observability UIs.
> ```

Nit: superfluous whitespace within "endpoints" word.
___
*[pkg/ccl/serverccl/tenant_grpc_test.go, line 44 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MgHngTh5Zm3QZr6-hJr:-MgHngTh5Zm3QZr6-hJs:b-c73vqe) ([raw file](https://github.com/cockroachdb/cockroach/blob/73090bd82ead796cdd220762b842f7338d38142c/pkg/ccl/serverccl/tenant_grpc_test.go#L44)):*
> ```Go
> 	server := testCluster.Server(0)
> 
> 	tenantID := roachpb.MakeTenantID(10)
> ```

You can use serverutils.TestTenantID() here instead of hardcoding the value 10.
___
*[pkg/server/server.go, line 1866 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MgHowJT3kE9qHqAUyju:-MgHowJT3kE9qHqAUyjv:beda752) ([raw file](https://github.com/cockroachdb/cockroach/blob/73090bd82ead796cdd220762b842f7338d38142c/pkg/server/server.go#L1866)):*
> ```Go
> // The function returns 3 arguments that are necessary to call
> // `RegisterGateway` which generated for each of your gRPC services
> // by grppc-gateway.
> ```

Nit, typp: grpc-gateway.


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
@rimadeodhar rebased!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-:-MgLUkvFCRMRFu21wPGS:b-b90snw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @dt, @knz, @rimadeodhar, and @stevendanna)

___
*[docs/generated/http/full.md, line 2803 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MgHmlFiDGV_SwXkzyw0:-MgLVIBaBHQVYgLNVL__:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/73090bd82ead796cdd220762b842f7338d38142c/docs/generated/http/full.md#L2803)):*
<details><summary><i>Previously, rimadeodhar (Rima Deodhar) wrote…</i></summary><blockquote>

Nit: Is this a typo, should it be serve fan-out requests?
</blockquote></details>

Done.
___
*[pkg/ccl/serverccl/tenant_grpc_test.go, line 32 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MgHnEok3i9VFAkgk9Ww:-MgLVJn51eIEFttxqeje:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/73090bd82ead796cdd220762b842f7338d38142c/pkg/ccl/serverccl/tenant_grpc_test.go#L32)):*
<details><summary><i>Previously, rimadeodhar (Rima Deodhar) wrote…</i></summary><blockquote>

Nit: superfluous whitespace within "endpoints" word.
</blockquote></details>

Done.
___
*[pkg/ccl/serverccl/tenant_grpc_test.go, line 44 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MgHngTh5Zm3QZr6-hJr:-MgLURiX7RjQ2kme_u8X:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/73090bd82ead796cdd220762b842f7338d38142c/pkg/ccl/serverccl/tenant_grpc_test.go#L44)):*
<details><summary><i>Previously, rimadeodhar (Rima Deodhar) wrote…</i></summary><blockquote>

You can use serverutils.TestTenantID() here instead of hardcoding the value 10.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
> Reviewed 67 of 67 files at r13.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66675)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @dt, @rimadeodhar, and @stevendanna)

___
*[pkg/rpc/auth_tenant.go, line 58 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfIM2HkA-twWeq19Cdt:-MgLl9Wo7ApZag9PcyPL:b-2cxry3) ([raw file](https://github.com/cockroachdb/cockroach/blob/11febcbd36a177ba89e7ff9668e2fee4681af785/pkg/rpc/auth_tenant.go#L58)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

> the error you observe suggests that a Statement RPC request is being sent in the (B) case somewhere in test TestTenantGRPCServices.  That request should not happen  - the test is doing something invalid. The problem is in that test and the fix is to change the test. How to prevent this test from sending this invalid request?

I don't think this is a test issue. The test dials the address of the tenant so it can't be accessing a (B) server. The issue is that tenant gRCP servers are configured in an identical manner to KV nodes. All gRPC servers are created using the same constructor `rpc.NewServer()` which adds these new auth interceptors for all requests that come from a tenant. When a tenant starts a gRPC server and accepts requests from other tenants, it opts-in in these authz routines as well.

Here are my options:

1. Make the `NewServer()` either configurable or aware if its being configured on a tenant node and disable the additional authz interceptors in those cases.
2. Make `authorize()` do something useful for tenant-tenant comms. It could, for instance, check that the client's tenantID matches the server tenantID. I don't think we explicitly check for that anywhere today.
3. Add a `// TODO` for more complex authz specification for tenant/tenant comms and move on

I'm taking option 2 as it provides us a way to retain common configuration while also restricting certain RPC calls to remain strictly between tenants when the client issuing the request is a tenant.

---

1. I've added a test `statements_test.go` in the `server` package that tests the (A) case. 

I added a 2 more test cases in my test file that issue a Statements RPC call (as opposed to an HTTP one) using one tenant's RPCContext to another (expecting success) and a request from tenant to KV (expecting failure)

2. Added comments
3. Added comments
</blockquote></details>

this is slightly better!


I still would like to express consternation at the fact that the RPC server is set up the same for tenant servers as KV nodes. This should never have happened... Mind filing up a follow-up issue? thanks


<!-- Sent from Reviewable.io -->

rimadeodhar reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Some minor nits around the TODOs but the code looks good. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-:-MgNCm4W03YpNaOUIqPo:b-vep25z)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @Azhng, @dhartunian, @dt, @rimadeodhar, and @stevendanna)

___
*[pkg/server/server.go, line 1867 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MgMEkmDDc03F4K_vgVV:-MgMEkmDDc03F4K_vgVW:b88gxdj) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2d58a4b9fa851aa18c4826d42feff1331816726/pkg/server/server.go#L1867)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
> 	// TODO(davidh): BEFORE MERGE: need guidance on how to cull the number
> 	// of contexts used here. Seems like we need a separate worker ctx at
> 	// least but maybe it can be crated in the method instead of passed
> 	// in
> ```
</details>

___
*[pkg/server/tenant.go, line 87 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MgMRQm1-3czuOCanMbh:-MgMRQm22BmzB-VE7UdN:b-66o9bg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2d58a4b9fa851aa18c4826d42feff1331816726/pkg/server/tenant.go#L87)):*
> ```
> 	// TODO(davidh): Do we need to force this to be false?
> 	baseCfg.SplitListenSQL = false
> ```

Seems like a comment that should be resolved prior to merge?
___
*[pkg/server/tenant.go, line 397 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MgMRug1ErZK36P6SI_z:-MgMRug1ErZK36P6SIa-:b-gw8tz8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2d58a4b9fa851aa18c4826d42feff1331816726/pkg/server/tenant.go#L397)):*
> ```Go
> 	// server to register against (but they'll never get RPCs at the time of
> 	// writing): the blob service and DistSQL.
> 	// TODO(davidh): anything to do below??? replace with server we added?
> ```

Ditto as above. Feels like the dummyRPC server should be replaced?
___
*[pkg/server/tenant_status.go, line 278 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MgNC31A8dCeA5xHT-vh:-MgNC31A8dCeA5xHT-vi:b-7x1n5s) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2d58a4b9fa851aa18c4826d42feff1331816726/pkg/server/tenant_status.go#L278)):*
<details><summary><i alt="summary">Quoted 6 lines of code…</i></summary>

> ```
> // iteratePods is based on the implementation of `iterateNodes in the
> // status server the two implementations have not been unified into one
> // because there are some deep differences since we use `InstanceInfo`
> // instead of `NodeID`. Since the eventual plan is to deprecate
> // `tenant_status.go` altogether, we're' leaving this code-as is.
> //
> ```
</details>


Nit: Missing period between the first two sentences and there seems be an extra apostrophe after the we're
___
*[pkg/server/serverpb/status.pb.go, line 0 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MgNCYR_29N6VIKRBrc3:-MgNCYR_29N6VIKRBrc4:bk0ygie) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2d58a4b9fa851aa18c4826d42feff1331816726/pkg/server/serverpb/status.pb.go#L0)):*
This file should probably be reverted?


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66675)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @Azhng, @dt, @knz, @rimadeodhar, and @stevendanna)

___
*[pkg/rpc/auth_tenant.go, line 58 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MfIM2HkA-twWeq19Cdt:-MgfSalb5UBxTOAcqNlM:bxt8z0c) ([raw file](https://github.com/cockroachdb/cockroach/blob/11febcbd36a177ba89e7ff9668e2fee4681af785/pkg/rpc/auth_tenant.go#L58)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

this is slightly better!


I still would like to express consternation at the fact that the RPC server is set up the same for tenant servers as KV nodes. This should never have happened... Mind filing up a follow-up issue? thanks

</blockquote></details>

Filed https://github.com/cockroachdb/cockroach/issues/68594
___
*[pkg/server/server.go, line 1867 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MgMEkmDDc03F4K_vgVV:-MgRFau--ABsuRhv4yOq:bb4fx24) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2d58a4b9fa851aa18c4826d42feff1331816726/pkg/server/server.go#L1867)):*
<details><summary><i>Previously, rimadeodhar (Rima Deodhar) wrote…</i></summary><blockquote>

> ```
> 	// TODO(davidh): BEFORE MERGE: need guidance on how to cull the number
> 	// of contexts used here. Seems like we need a separate worker ctx at
> 	// least but maybe it can be crated in the method instead of passed
> 	// in
> ```

</blockquote></details>

Removed. I found precedent in `server.go` for passing the additional `workersCtx` and `ambientCtx` contexts to helper functions.
___
*[pkg/server/tenant.go, line 87 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MgMRQm1-3czuOCanMbh:-MgRGRhI2gSjSntnCn5F:b-wnfmk9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2d58a4b9fa851aa18c4826d42feff1331816726/pkg/server/tenant.go#L87)):*
<details><summary><i>Previously, rimadeodhar (Rima Deodhar) wrote…</i></summary><blockquote>

> ```
> 	// TODO(davidh): Do we need to force this to be false?
> 	baseCfg.SplitListenSQL = false
> ```

Seems like a comment that should be resolved prior to merge?
</blockquote></details>

Done. Removed.
___
*[pkg/server/tenant.go, line 397 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MgMRug1ErZK36P6SI_z:-MgfJG018Dp1hHF4UAsD:b-e68hum) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2d58a4b9fa851aa18c4826d42feff1331816726/pkg/server/tenant.go#L397)):*
<details><summary><i>Previously, rimadeodhar (Rima Deodhar) wrote…</i></summary><blockquote>

Ditto as above. Feels like the dummyRPC server should be replaced?
</blockquote></details>

In this case I was unable to easily make the change without modifying much more about the startup order of operations which I want to avoid.

The `dummyRPC` is being used for the `sqlServerOptionalKVArgs` which are described as only being important for SQL that runs on the KV nodes so I think in this case it's fine to leave as-is.
___
*[pkg/server/tenant_status.go, line 278 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MgNC31A8dCeA5xHT-vh:-MgfJfSf731FoNrbhA7E:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2d58a4b9fa851aa18c4826d42feff1331816726/pkg/server/tenant_status.go#L278)):*
<details><summary><i>Previously, rimadeodhar (Rima Deodhar) wrote…</i></summary><blockquote>

> ```
> // iteratePods is based on the implementation of `iterateNodes in the
> // status server the two implementations have not been unified into one
> // because there are some deep differences since we use `InstanceInfo`
> // instead of `NodeID`. Since the eventual plan is to deprecate
> // `tenant_status.go` altogether, we're' leaving this code-as is.
> //
> ```

Nit: Missing period between the first two sentences and there seems be an extra apostrophe after the we're
</blockquote></details>

Done.
___
*[pkg/server/serverpb/status.pb.go, line  at r13](https://reviewable.io/reviews/cockroachdb/cockroach/66675#-MgNCYR_29N6VIKRBrc3:-MgfLBnD-SgyjuzB91Mn:b-8moobb) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2d58a4b9fa851aa18c4826d42feff1331816726/pkg/server/serverpb/status.pb.go#L)):*
<details><summary><i>Previously, rimadeodhar (Rima Deodhar) wrote…</i></summary><blockquote>

This file should probably be reverted?
</blockquote></details>

I added a comment in `status.proto` hence the change in here. The diff got weird after rebase.


<!-- Sent from Reviewable.io -->

dhartunian commented:
bors r+
